APDS9922 - remove unnecessary spinlock, consolidate i2c read/write and fix incorrect variable declaration and other type errors

TimJTi opened the pull request:
## Summary
If both the ambient light sensor AND proximity sensor are enabled (and therefore registered as two independent devices) spinlock is required to ensure I2C bus transactions don't conflict

A simple #warning  has been added to notify the user

## Impact
None - hopefuly will stop others wasting hours wondering why their app crashes!

## Testing
Custom board with SAMA5D27C-D1M with 2x apds9922 devices fitted

acassis commented:
Hi @TimJTi according with Kconfig help:
```
config SPINLOCK
        bool "Support Spinlocks"
        default n
        ---help---
                Enables support for spinlocks.  Spinlocks are used primarily for
                synchronization in SMP configurations but are available for general
                synchronization between CPUs.  Use in a single CPU configuration would
                most likely be fatal.
               config SPINLOCK
        bool "Support Spinlocks"
        default n
        ---help---
                Enables support for spinlocks.  Spinlocks are used primarily for
                synchronization in SMP configurations but are available for general
                synchronization between CPUs.  Use in a single CPU configuration would
                most likely be fatal.
```

Shouldn't if be:

```
#if defined(CONFIG_SMP) && !defined(CONFIG_SPINLOCK)
```

?

acassis commented on drivers/sensors/apds9922.c:
```diff
@@ -48,6 +48,10 @@
  ****************************************************************************/
 
+#ifndef CONFIG_SPINLOCK
+#  warning Spinlock should be enabled if light AND proximity detection wanted
```
```suggestion
#  warning Spinlock (CONFIG_SPINLOCK) should be enabled if light AND proximity detection wanted
```

TimJTi commented:
> Shouldn't if be:
> 
> ```
> #if defined(CONFIG_SMP) && !defined(CONFIG_SPINLOCK)
> ```

This is not to do with SMP - it is to do with one device on a single I2C bus performing two different functions with 2 different registered drivers (/dev/als0 and /dev/ps0). My configs certainly don't have CONFIG_SMP set, so maybe I shouldn't have used a spinlock to protect the I2C read/writes when I wrote this driver?

TimJTi commented:
> Shouldn't it be:
> 
> ```
> #if defined(CONFIG_SMP) && !defined(CONFIG_SPINLOCK)
> ```
> 

@acassis - I think I used spinlock because someone suggested it when I was adding support for the ACT8945A which is both a power supply device and battery charger device, so the same issue, potentially.

Is this something a mutex should actually be used for? Happy to admit it's on the edges of my understanding of NuttX/RTOS' at this time :-o

pkarashchenko commented:
Sorry, I'm not fully aware of the problem that is intended to be solved with spin locks here so will start to ask dumb questions:
What is the actual issue that requires spin locks to solve it? I just took a brief look into the code and see things like
```
static int apds9922_i2c_read(FAR struct apds9922_dev_s *priv,
                             uint8_t const regaddr,
                             FAR uint8_t *regval, int len)
{
  struct i2c_config_s config;
  int                 ret;
  irqstate_t          flags;

  DEBUGASSERT(priv);

  /* Set up the I2C configuration */

  config.frequency = CONFIG_APDS9922_I2C_FREQUENCY;
  config.address   = priv->config->i2c_addr;
  config.addrlen   = 7;

  /* Write the register address to read from */

  flags = spin_lock_irqsave(NULL);
  ret = i2c_write(priv->config->i2c, &config, &regaddr, 1);
  spin_unlock_irqrestore(NULL, flags);
  if (ret < 0)
    {
      snerr ("i2c_write failed: %d\n", ret);
      return ret;
    }

  /* Read "len" bytes from regaddr */

  flags = spin_lock_irqsave(NULL);
  ret = i2c_read(priv->config->i2c, &config, regval, len);
  spin_unlock_irqrestore(NULL, flags);
  if (ret < 0)
    {
      snerr ("i2c_read failed: %d\n", ret);
      return ret;
    }

  return OK;
}
```
so the obvious question is maybe we can use `i2c_writeread` instead of `i2c_write + i2c_read` like
```
static int apds9922_i2c_read(FAR struct apds9922_dev_s *priv,
                             uint8_t const regaddr,
                             FAR uint8_t *regval, int len)
{
  struct i2c_config_s config;
  int                 ret;

  DEBUGASSERT(priv);

  /* Set up the I2C configuration */

  config.frequency = CONFIG_APDS9922_I2C_FREQUENCY;
  config.address   = priv->config->i2c_addr;
  config.addrlen   = 7;

  /* Write the register address to read from and read "len" bytes from regaddr */

  ret = i2c_writeread(priv->config->i2c, &config, &regaddr, 1, regval, len);
  if (ret < 0)
    {
      snerr ("i2c_writeread failed: %d\n", ret);
      return ret;
    }

  return OK;
}
```
and get rid of spin locks at all?

TimJTi commented:
That makes sense. I'm learning all the time and I used spinlock as it had been suggested here for a similar purpose.

Will try tomorrow and report back!

TimJTi commented:
@pkarashchenko
> What is the actual issue that requires spin locks to solve it?

The issue is that the APDS9922 is both an ambient light sensor and proximity sensor. These are two different functions, registered as two different devices.

What I found is that there were problems (i.e. a crash) because the two different drivers canobviously access the same device on the same i2c bus at random times; not surprisingly, of course.

Spinlock seemed to be appropriate simply based on a suggestion from @acassis via the mailing list in relation to a similar issue with the ACT8945A - another dual function device (battery charger and cpu regulator):

> But you will need to use spinlock when accessing the I2C bus inside these drivers to avoid both drivers trying to use it while a transfer is in progress.

Read/write.ioctl/etc use a common mutex to avoid problems but the collecting of data is autonomous, with different data-ready timing for each sensor (i.e. they are asynchronous).

If spinlock isn't the answer then maybe I should make use of the mutex already included in the apds9922 device struct to protect the i2c read/write/readwrite functions?

> maybe we can use i2c_writeread instead of i2c_write + i2c_read

That is obviously much neater than a call to write followed by a call to a  read so I can certainly add that and make use of it to make the code neater, but I don't think it relates to the I2C clashing issue itself?

TimJTi commented:
I have removed the spinlocks, changed it to use `i2c_writeread,` rationalised some other stuff and fixed a few type mismatches and it seems to be behaving OK.

All the ioctl, read, and write "upper driver" calls are protected by a mutex so I think I put the spinlock there as I thought I should, and whatever problem I was seeing back then was most likely something else. Fingers crossed - but I will be using this driver thoroughly in the coming months and'll soon find any problems!

I will push my changes soon.

pkarashchenko commented on drivers/sensors/apds9922.c:
```diff
@@ -444,19 +438,19 @@ static void apds9922_worker(FAR void *arg)
 
       if ((priv->ps_setup.notify != PS_PROXIMITY_DATA_ONLY) &&
-         (priv->ps_data->close != (status & PS_LOGIC_STATUS)))
+         (priv->ps_data.close != (status & PS_LOGIC_STATUS)))
```
```suggestion
          (priv->ps_data.close != (status & PS_LOGIC_STATUS)))
```

pkarashchenko reviewed (approved):
Looks much better now

pkarashchenko commented on drivers/sensors/apds9922.c:
```diff
@@ -254,7 +253,7 @@ struct apds9922_dev_s
   struct apds9922_ps_setup_s   ps_setup;  /* Device PS config           */
   int                          als;       /* ALS data                   */
-  FAR struct apds9922_ps_data  *ps_data;  /* PS data                    */
+  FAR struct apds9922_ps_data  ps_data;   /* PS data                    */
```
```suggestion
  struct apds9922_ps_data  ps_data;       /* PS data                    */
```

pkarashchenko commented:
Please squash changes into 1 commit

TimJTi commented:
> Please squash changes into 1 commit

Yep - just waiting for CI checks to finish, just in case!

TimJTi commented:
> I have removed the spinlocks...
Looking back at the comment from @acassis regarding spinlock use it was the ACT8945A which will have two completely different drivers - one for regulator, one for battery charger. Spinlock was never appropriate here.

But that leaves the question of what is the correct method as and when I add the battery charger driver for the ACT8945A? Same chip, same I2C bus, same I2C address. Is spinlock the right thing there? Or is there a higher-level mutex or something? Spinlock "feels" right though?

Obviously a question for another time/place unless anyone following this has a view?

pkarashchenko commented:
I will take a closer loop to your question. At the first glance the spinlock is an overkill and I2C mutex protection should be enough if we can ensure that a single transaction is not interrupted by other users, just like we did with the write-read interface.
We need an additional level of protection only if accessing desired functionality requires multiple i2c (write-read-write-read for example) transactions that can't be interrupted by other users of the same device over i2c.

TimJTi commented:
> I will take a closer loop to your question. At the first glance the spinlock is an overkill and I2C mutex protection should be enough if we can ensure that a single transaction is not interrupted by other users, just like we did with the write-read interface. We need an additional level of protection only if accessing desired functionality requires multiple i2c (write-read-write-read for example) transactions that can't be interrupted by other users of the same device over i2c.

From a quick look (and apologies that it was only a quick look) the low level I2C transaction are not protected, but within a given driver, the upper half open/close/ioctl etc within the lower half driver are mutex protected. 

With the ACT8945A there will be two completely independent sets of driver code with, I think, a potential for I2C clashes unless the read/write sequences are protected by something at a higher level?

pkarashchenko commented:
But there is a lock at the arch level

TimJTi commented:
Good! I just took the dev list comment at face value. Thanks for checking.

mu578 commented:
If you had some kind of non-atomic read-write operations within a same context you could use a spinlock, i.e acting like a constant monitoring timed switch, but here you have two distinct drivers + I don't think a bit of latency is an issue for you, until one transaction is mutex locked any other attempt to access BUS will be postponed.
